[PATCH] [Xtensa] Implement FrameLowering methods and stack operation lowering.

andreisfr opened the pull request:
Implement emitPrologue/emitEpilogue methods, determine/spill/restore callee saved registers functionality with test. Also implement lowering of the DYNAMIC_STACKALLOC/STACKSAVE/STACKRESTORE stack operations with tests.

andreisfr committed:
[PATCH] [Xtensa] Implement FrameLowering methods and stack operations lowering.

Implement emitPrologue/emitEpilogue methods, determine/spill/restore callee
saved registers functionality with test. Also implement lowering of the
DYNAMIC_STACKALLOC/STACKSAVE/STACKRESTORE stack operations with tests.

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -33,10 +33,158 @@ bool XtensaFrameLowering::hasFP(const MachineFunction &MF) const {
+                                       MachineBasicBlock &MBB) const {
+  assert(&MBB == &MF.front() && "Shrink-wrapping not yet implemented");
+  MachineFrameInfo &MFI = MF.getFrameInfo();
+  const XtensaRegisterInfo *RegInfo = static_cast<const XtensaRegisterInfo *>(
```
getSubtarget<XtensaSubtarget>().get...

andreisfr replied:
Fxied

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -33,10 +33,158 @@ bool XtensaFrameLowering::hasFP(const MachineFunction &MF) const {
+  MachineFrameInfo &MFI = MF.getFrameInfo();
+  const XtensaRegisterInfo *RegInfo = static_cast<const XtensaRegisterInfo *>(
+      MF.getSubtarget().getRegisterInfo());
+  const XtensaInstrInfo &TII =
```
Ditto, also just save subtarget to a variable

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -33,10 +33,158 @@ bool XtensaFrameLowering::hasFP(const MachineFunction &MF) const {
+      *static_cast<const XtensaInstrInfo *>(MF.getSubtarget().getInstrInfo());
+  MachineBasicBlock::iterator MBBI = MBB.begin();
+  DebugLoc DL = MBBI != MBB.end() ? MBBI->getDebugLoc() : DebugLoc();
+  unsigned SP = Xtensa::SP;
```
Use MCRegister

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -33,10 +33,158 @@ bool XtensaFrameLowering::hasFP(const MachineFunction &MF) const {
+                                       MachineBasicBlock &MBB) const {
+  MachineBasicBlock::iterator MBBI = MBB.getLastNonDebugInstr();
+  MachineFrameInfo &MFI = MF.getFrameInfo();
+  const XtensaRegisterInfo *RegInfo = static_cast<const XtensaRegisterInfo *>(
```
Same as other function

andreisfr replied:
Fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -33,10 +33,158 @@ bool XtensaFrameLowering::hasFP(const MachineFunction &MF) const {
+    // Find the first instruction that restores a callee-saved register.
+    MachineBasicBlock::iterator I = MBBI;
+
+    for (unsigned i = 0; i < MFI.getCalleeSavedInfo().size(); ++i)
```
I don't see how incrementing the instruction iteration based on this makes sense

andreisfr replied:
Thank you very much for review. I've made some code improvements (added more comments and loop is  changed), I hope that I understand correctly your comment. Actually I used the same approach as in Mips architecture https://github.com/llvm/llvm-project/blob/57c10fa564af44a5b236bc642c540d715b04448c/llvm/lib/Target/Mips/MipsSEFrameLowering.cpp#L711 , because it is suitable for Xtensa.

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -57,3 +205,31 @@ MachineBasicBlock::iterator XtensaFrameLowering::eliminateCallFramePseudoInstr(
+                                               BitVector &SavedRegs,
+                                               RegScavenger *RS) const {
+  MachineFrameInfo &MFI = MF.getFrameInfo();
+  const XtensaRegisterInfo *RegInfo = static_cast<const XtensaRegisterInfo *>(
```
Ditto

andreisfr replied:
fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -57,3 +205,31 @@ MachineBasicBlock::iterator XtensaFrameLowering::eliminateCallFramePseudoInstr(
+    return;
+
+  const TargetRegisterClass &RC = Xtensa::ARRegClass;
+  const TargetRegisterInfo *TRI = MF.getSubtarget().getRegisterInfo();
```
Ditto

andreisfr replied:
fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -534,13 +540,56 @@ SDValue XtensaTargetLowering::LowerConstantPool(ConstantPoolSDNode *CP,
+
+SDValue XtensaTargetLowering::LowerSTACKRESTORE(SDValue Op,
+                                                SelectionDAG &DAG) const {
+  unsigned SP = Xtensa::SP;
```
No need for the variable

andreisfr replied:
fixed

andreisfr committed:
[PATCH] [Xtensa] Minor fixes in FrameLowering and ISelLowering.

Remove redundant code from XtensaFrameLowering and
XtensaISelLowering, add comments.

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -33,10 +34,153 @@ bool XtensaFrameLowering::hasFP(const MachineFunction &MF) const {
+
+  const std::vector<CalleeSavedInfo> &CSI = MFI.getCalleeSavedInfo();
+
+  if (CSI.size()) {
```
!CSI.empty()

andreisfr replied:
fixed

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -33,10 +34,153 @@ bool XtensaFrameLowering::hasFP(const MachineFunction &MF) const {
+  const std::vector<CalleeSavedInfo> &CSI = MFI.getCalleeSavedInfo();
+
+  if (CSI.size()) {
+    // Find the instruction past the last instruction that saves a
+    // callee-saved register to the stack.
+    for (unsigned i = 0; i < CSI.size(); ++i)
+      ++MBBI;
```
This seems like a dubious way to guarantee this. Can you at least assert each instruction is a spill of the appropriate register?

andreisfr replied:
I added check functions for store/restore instructions, PTAL.

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -33,10 +34,153 @@ bool XtensaFrameLowering::hasFP(const MachineFunction &MF) const {
+    // so we should find first instruction of the sequence.
+    MachineBasicBlock::iterator I = MBBI;
+
+    // Find the first instruction at the end that restores a callee-saved
+    // register.
+    for (auto &Info : MFI.getCalleeSavedInfo())
+      --I;
```
Same as above, can this at least assert the instructions are the expected ones?

andreisfr replied:
I added check functions

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -51,9 +195,34 @@ MachineBasicBlock::iterator XtensaFrameLowering::eliminateCallFramePseudoInstr(
+  Align Alignment = TRI->getSpillAlign(RC);
+  int FI = MF.getFrameInfo().CreateStackObject(Size, Alignment, false);
+
+  RS->addScavengingFrameIndex(FI);
```
I'm not sure it's safe to do this in determineCalleeSaves. Can you create the emergency scavenging index in processFunctionBeforeFrameFinalized instead?

andreisfr replied:
I moved frame index scavenging to processFunctionBeforeFrameFinalized

andreisfr committed:
[PATCH] [Xtensa] Check save/restore instructions in emitPrologue and emitEpilogue.

Check whether store/restore instruciton for callee-saved registers
are placed correctly in emitProlgue and emitEpilogue. Also move
frame index scavenging from determineCalleeSaves function to
processFunctionBeforeFrameFinalized function.

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -33,6 +33,70 @@ bool XtensaFrameLowering::hasFP(const MachineFunction &MF) const {
+#ifndef NDEBUG
+/* Check whether instruction I stores some callee-saved register from CSI
+ */
+static bool checkStoreInstruction(MachineBasicBlock::iterator I,
+                                  const std::vector<CalleeSavedInfo> &CSI) {
```
Name is too general. Also, this is way more code than I expected for this assert. Don't you expect these instructions to appear in order? 

Can you use TII->isStoreToStackSlot and isCopyInstr to avoid special casing these specific opcodes?

andreisfr replied:
Yes, I will implement then isStoreToStackSlot /isLoadFromStackSlot functions. Current PrologueEpilogueInsterter implementation places stores in direct order:
.. for (const CalleeSavedInfo &CS : CSI)..
and uses reverse order for placement loads at the end of basic block. So, we can assume that we always have such order and create redundant version which checks instructions. Is it acceptable?

arsenm replied:
If you're relying on seeing these exact instructions, it's OK to rely on the order. The point of the assert was to verify the instructions exactly match what you expected, which includes the order

andreisfr committed:
[PATCH] [Xtensa] Simplify calle-saved register checking.

arsenm commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:
```diff
@@ -48,6 +48,30 @@ XtensaInstrInfo::XtensaInstrInfo(const XtensaSubtarget &STI)
+      return MI.getOperand(0).getReg();
+    }
+  }
+  return 0;
```
return Register()

arsenm commented on llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:
```diff
@@ -48,6 +48,30 @@ XtensaInstrInfo::XtensaInstrInfo(const XtensaSubtarget &STI)
+      return MI.getOperand(0).getReg();
+    }
+  }
+  return 0;
```
Register()

andreisfr committed:
[PATCH] [Xtensa] Minor changes in callee-saved registers checks

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -33,10 +34,194 @@ bool XtensaFrameLowering::hasFP(const MachineFunction &MF) const {
+#ifndef NDEBUG
+      // Checking that the instruction is exactly as expected
+      bool IsStoreInst = false;
+      if ((MBBI->getOpcode() == TargetOpcode::COPY) && Info.isSpilledToReg()) {
```
don't need these braces

andreisfr replied:
I corrected "if-condition", is it ok? Also, sorry. I changed loop structure a bit with callee-saved register checks to avoid compiler messages like "unused variable" and fix variable name typo. Could you PTAL? Thanks

andreisfr committed:
[PATCH] [Xtensa] Fix compile warnings anf variable name typo.

MaskRay commented:
I fixed two warnings in dd82fd4744397e0510c8204f1a6031441e21858e. You might want to build llvm-project with a not-too-ancient clang to ensure that there is no such warning for future patches:)

`[PATCH] ` in the title is not needed.

andreisfr commented:
> I fixed two warnings in [dd82fd4](https://github.com/llvm/llvm-project/commit/dd82fd4744397e0510c8204f1a6031441e21858e). You might want to build llvm-project with a not-too-ancient clang to ensure that there is no such warning for future patches:)
> 
> `[PATCH] ` in the title is not needed. The majority of patches omit the trailing full stop in the title.
Thank you very much, I'll take this into account.
